src/openforms/tests/e2e/input_validation_base.py

```diff
@@ -0,0 +1,145 @@
+from typing import TypedDict
+
+from django.test import override_settings, tag
+from django.urls import reverse
+
+from asgiref.sync import async_to_sync
+from furl import furl
+from playwright.async_api import Page, expect
+from rest_framework import status
+from typing_extensions import NotRequired, Unpack
+
+from openforms.formio.typing import Component
+from openforms.forms.models import Form
+from openforms.forms.tests.factories import (
+    FormDefinitionFactory,
+    FormFactory,
+    FormStepFactory,
+)
+from openforms.submissions.tests.factories import SubmissionFactory
+from openforms.submissions.tests.mixins import SubmissionsMixin
+from openforms.typing import JSONValue
+
+from .base import E2ETestCase, browser_page
+
+
+def create_form(component: Component) -> Form:
+    _test_name = f"{component['type']} - {component['label']}"[:50]
+    form_definition = FormDefinitionFactory.create(
+        name=_test_name,
+        configuration={"components": [component]},
+    )
+    form = FormFactory.create(
+        name=_test_name,
+        product=None,
+        category=None,
+        registration_backend=None,
+        translation_enabled=False,  # force Dutch
+        ask_privacy_consent=False,
+        ask_statement_of_truth=False,
+    )
+    FormStepFactory.create(
+        form=form, form_definition=form_definition, next_text="Volgende"
+    )
+    return form
+
+
+class ValidationKwargs(TypedDict):
+    api_value: NotRequired[JSONValue]
+
+
+@tag("e2e-validation")
+# allow all origins, since we don't know exactly the generated live server port number
+@override_settings(CORS_ALLOW_ALL_ORIGINS=True)
+class ValidationsTestCase(SubmissionsMixin, E2ETestCase):
+
+    def assertValidationIsAligned(
+        self,
+        component: Component,
+        ui_input: str | int | float,
+        expected_ui_error: str,
+        **kwargs: Unpack[ValidationKwargs],
```

Comment:
Funny alternative to not use an Unset default value

Reply:
I tried using a singleton but that's horrendous with type checking :(